src/lifetime_free.rs

```diff
@@ -0,0 +1,81 @@
+/// Marker trait for types that do not contain any lifetime parameters. Such
+/// types are safe to cast from non-static type parameters if their types are
+/// equal.
+///
+/// This trait is used by [`cast!`] to determine what casts are legal on values
+/// without a `'static` type constraint.
+///
+/// # Safety
+///
+/// When implementing this trait for a type, you must ensure that the type is
+/// free of any lifetime parameters. Failure to meet **all** of the requirements
+/// below may result in undefined behavior.
+///
+/// - The type must be `'static`.
+/// - The type must be free of lifetime parameters. In other words, the type
+///   must be an "owned" type and not contain *any* lifetime parameters.
+/// - All contained fields must also be `LifetimeFree`.
+///
+/// # Examples
+///
+/// ```
+/// use castaway::LifetimeFree;
+///
+/// struct Container<T>(T);
+///
+/// // UNDEFINED BEHAVIOR!!
+/// // unsafe impl LifetimeFree for Container<&'static str> {}
+///
+/// // UNDEFINED BEHAVIOR!!
+/// // unsafe impl<T> LifetimeFree for Container<T> {}
+///
+/// // This is safe.
+/// unsafe impl<T: LifetimeFree> LifetimeFree for Container<T> {}
+///
+/// struct PlainOldData {
+///     foo: u8,
+///     bar: bool,
+/// }
+///
+/// // This is also safe, since all fields are known to be `LifetimeFree`.
+/// unsafe impl LifetimeFree for PlainOldData {}
+/// ```
+pub unsafe trait LifetimeFree: 'static {}
+
+unsafe impl LifetimeFree for () {}
+unsafe impl LifetimeFree for bool {}
+unsafe impl LifetimeFree for u8 {}
+unsafe impl LifetimeFree for u16 {}
+unsafe impl LifetimeFree for u32 {}
+unsafe impl LifetimeFree for u64 {}
+unsafe impl LifetimeFree for u128 {}
+unsafe impl LifetimeFree for usize {}
+unsafe impl LifetimeFree for i8 {}
+unsafe impl LifetimeFree for i16 {}
+unsafe impl LifetimeFree for i32 {}
+unsafe impl LifetimeFree for i64 {}
+unsafe impl LifetimeFree for i128 {}
+unsafe impl LifetimeFree for isize {}
+unsafe impl LifetimeFree for f32 {}
+unsafe impl LifetimeFree for f64 {}
+unsafe impl LifetimeFree for char {}
+unsafe impl LifetimeFree for str {}
```

Comment:
Can we also implement it for [NonZero* types in core](https://doc.rust-lang.org/stable/std/num/index.html)?

Reply:
Yep, that makes sense.